### ISSUE TITLE
Accessing array elements by key

### ISSUE THREAD

**Issue by mdehoon** [numpy/numpy#15453]
I am looking for an easy way to access array elements by (string) key (as well as by index).
Suppose I have an array like this:
```
x = array([[0, 4, 9, 1],
           [1, 3, 9, 1],
           [3, 5, 6, 2],
           [6, 2, 7, 5]])
```
I am looking for way to specify a set of keys (for example `('A', 'C', 'G', 'T')`) that can be used as an alias for an index. So `x['A', 'C']`, `x[0,'C']`, `x['A', 1]`, and `x[0,1]` all return the value 4; `x['G', :]` is the same as `x[2, :]`, and so on. I know that this can be achieved by subclassing a numpy array and overriding `__getitem__` and `__setitem__`, but subclassing gets complicated very quickly.

I understand that a pandas `DataFrame` has this capability, but as far as I understand it does not subclass from a numpy array, and therefore (unlike numpy arrays) does not provide the buffer protocol in C code.

**Comment by eric-wieser** [comment]
@shoyer, does pydata/xarray work well here?

I know it's hacky, but have you considered:
```
A, C, G, T = range(3)
x[A, C]
```

**Comment by mdehoon** [comment]
@eric-wieser  Never thought of that. But yes that would be too hacky. This is for an open-source project, and in their scripts users may be using the `A`, `C`, `G`, `T` variables for other things.

**Comment by shoyer** [comment]
> I understand that a pandas `DataFrame` has this capability, but as far as I understand it does not subclass from a numpy array, and therefore (unlike numpy arrays) does not provide the buffer protocol in C code.

True, but you can convert a pandas array into a NumPy array (typically without a copy), which satisfies the buffer interface. You can do the same for xarray.

(I'm actually not quite sure why you can't define the buffer interface for types written in Python, but indeed that seems to be the case.)
